8305753: Allow JIT compilation for -Xshare:dump

iklam opened the pull request:
This RFE makes it possible for the user to enable the JIT compiler during the CDS dump by specifying `-Xshare:dump -Xmixed`

By default, `-Xshare:dump` runs with `-Xint`. This maintains the previous behavior.

Which mode to use?

- Dumping in `-Xint` mode is faster when the classlist is small. `-Xint` is also required for deterministic dumping (compiler threads can change the order of the archived object).

- However, dumping in `-Xmixed` mode is faster when the class list is larger. We also need to run the compiler in Leyden so that we can generate the CDS archive and AOT code at the same time.


Tested with tiers 1-4 and builds-tier5.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8305753](https://bugs.openjdk.org/browse/JDK-8305753): Allow JIT compilation for -Xshare:dump (**Enhancement** - P4)


### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**)
 * [Matias Saavedra Silva](https://openjdk.org/census#matsaave) (@matias9927 - Committer) ⚠️ Review applies to [0931637e](https://git.openjdk.org/jdk/pull/16233/files/0931637e601f57c17f5793facfca984d2a3018d4)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to [0931637e](https://git.openjdk.org/jdk/pull/16233/files/0931637e601f57c17f5793facfca984d2a3018d4)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16233/head:pull/16233` \
`$ git checkout pull/16233`

Update a local copy of the PR: \
`$ git checkout pull/16233` \
`$ git pull https://git.openjdk.org/jdk.git pull/16233/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16233`

View PR using the GUI difftool: \
`$ git pr show -t 16233`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16233.diff">https://git.openjdk.org/jdk/pull/16233.diff</a>

</details>


### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16233#issuecomment-1767690589)

dholmes-ora reviewed (commented):
Is the current use of `-Xint` documented? I'm not seeing how you are communicating this change to the user. To be honest this seems like it is aimed at Leyden, and as such probably belongs over there rather than in mainline.

matias9927 commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -3030,14 +3036,18 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
+    } else if (_mode == _comp) {
+      // -Xcomp may use excessive CPU for the test tiers. Also, -Xshare:dump runs a small and fixed set of
+      // Java code, so there's not much benefit in running -Xcomp.
+      log_info(cds)("reduced -Xcomp to -Xmixed");
```
Is it preferable to reduce the user-selected option over throwing an error for incompatible options?

iklam replied:
We have many test jobs that add `-Xcomp` to every test case, so throwing an error will cause these tests to fail.

If we don't reduce the strength to `-Xmixed`, then many of these tests will fail due to timeout, forcing us to change the timeout value in these tests.

So I think changing to `-Xmixed` is a good compromise to do some testing with the JIT during `-Xshare:dump` without being intrusive.

matias9927 replied:
That makes perfect sense! I am still concerned about the idea taking a user option and reducing it to an option they didn't ask for... I will trust your judgement

matias9927 commented on test/hotspot/jtreg/runtime/cds/DeterministicDump.java:
```diff
@@ -79,6 +79,7 @@ static String dump(ArrayList<String> args, String... more) throws Exception {
         String archiveName = logName + ".jsa";
         String mapName = logName + ".map";
         CDSOptions opts = (new CDSOptions())
+            .addPrefix("-Xint") // Override any -Xmixed/-Xcomp flags from jtreg -vmoptions
```
I assume this issue could pop up in newer tests as well. Is there a place where you can specify that `-Xint` needs to be explicitly added when doing a static dump?

iklam replied:
-Xint is necessary only when we want to test for deterministic dumps. We don't have many such test cases.

matias9927 commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -3030,14 +3036,18 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
+      // -Xcomp may use excessive CPU for the test tiers. Also, -Xshare:dump runs a small and fixed set of
+      // Java code, so there's not much benefit in running -Xcomp.
+      log_info(cds)("reduced -Xcomp to -Xmixed");
+      set_mode_flags(_mixed);
```
Missing an else if case for `_mode==_mixed` ?

iklam replied:
That's intentional: if the user has explicitly chosen -Xint or -Xmixed, we don't override that.

matias9927 reviewed (commented):
Looks good but I have a few questions

calvinccheung reviewed (approved):
Changes seem fine.
Just a suggestion on the log message.
Also, please file a doc subtask if you think it needs to be documented in the release note.

calvinccheung commented on src/hotspot/share/runtime/arguments.cpp:
```diff
@@ -3030,14 +3036,18 @@ jint Arguments::finalize_vm_init_args(bool patch_mod_javabase) {
+    } else if (_mode == _comp) {
+      // -Xcomp may use excessive CPU for the test tiers. Also, -Xshare:dump runs a small and fixed set of
+      // Java code, so there's not much benefit in running -Xcomp.
+      log_info(cds)("reduced -Xcomp to -Xmixed");
```
I'd suggest the following log message:

`log_info(cds)("reduced -Xcomp to -Xmixed for static dumping");`

iklam commented:
> Changes seem fine. Just a suggestion on the log message. Also, please file a doc subtask if you think it needs to be documented in the release note.

I updated the log message as you suggested. I also added the label release-note=yes to the JBS issue.

dholmes-ora commented:
For this to be useful in mainline, per the example given, then users have to be very aware of its existence. A Release Note is not sufficient, this needs to be documented in the Java command reference (manpage) as well. Thanks.

matias9927 reviewed (approved):
Thanks for the clarification, this looks good!

iklam commented:
> For this to be useful in mainline, per the example given, then users have to be very aware of its existence. A Release Note is not sufficient, this needs to be documented in the Java command reference (manpage) as well. Thanks.

Hi David, I added documentation into the java.1 man page.

dholmes-ora reviewed (approved):
Manpage update looks good. Thanks

iklam commented:
Thanks @matias9927 @calvinccheung @dholmes-ora for the review
/integrate
